# Абстрактное и конкретное

### (для учеников младших классов)

Абстрактное и конкретное - это типы утверждений или высказываний, они представляют собой две стороны одной медали. Абстрактное и конкретное никогда не встречаются в чистом виде, любое высказывание может быть более или менее абстрактным, или, другими словами, менее или более конкретным. Проще всего понять смысл этих терминов через примеры. Сначала дадим их общее описание.

<b>Абстрагирование</b> - процесс отбрасывания деталей, которые считаются ненужными в данном контексте.<br>
<b>Конкретизация</b> - противоположный процесс привлечения дополнительных деталей, которые считаются нужными в данном контексте.

#### Задача 1

Рассмотрим <i>первую</i> задачу: у Пети есть 3 яблока, а у Васи - 5 яблок. Сколько всего яблок у мальчиков? Задачка решается легко: у мальчиков вместе 8 яблок.

#### Задача 2

А теперь применим <b>абстрагирование</b>: сочтем, что вместо яблок можно рассматривать любые другие предметы. Тогда задачу можно записать короче: 3 + 5 = 8. Это <i>вторая</i> задача, она абстрактнее первой, в ней отсутствует деталь (тип предметов, количества которых мы складываем - яблоки).

Главное достижение абстракции - общность. Теперь мы можем дегко решать аналогичные задачи с другим конкретным содержанием. Кроме того, такую абстрактную задачу можно записать и решить на компьютерном языке программирования, как показано в следующей строке.

In [1]:
3 + 5

8

#### Задача 3

А теперь ко второй задаче применим <b>конкретизацию</b>: за первый час ученик решил 3 примера из домашнего задания, а за второй час - 5 примеров. Сколько всего примеров решил ученик?

Эта задача конкретнее второй, в ней складываются не абстрактные числа, а количества конкретных предметов - примеров из домашнего задания.

Ясно, что количество решенных примеров равно 8, а для решения такой задачи можно применять решение более абстрактной задачи 2, которую мы решали раньше.

#### Задача 4

А теперь конкретизируем исходную задачу 1. Если нам требуется вычислить не количество яблок у мальчиков, а их вес, то необходимо привлечь дополнительные детали - вес каждого яблока, то есть, провести <b>конкретизацию</b>.

После измерения веса получаем следующие значения: яблоки Пети весят 150г, 220г и 170г, а яблоки Васи - 180г, 190г, 230г, 240г и 120г.

Для вычисления суммарного веса используем компьютер, получаем для веса Петиных яблок величину

In [2]:
150 + 220 + 170

540

а для веса васиных яблок - величину

In [3]:
180 + 190 + 230 + 240 + 120

960

Вместе яблоки весят

In [4]:
540 + 960

1500

1500 грамм, то есть, полтора килограмма.

Следующий вопрос, который может нас заинтересовать: сколько стоят все эти яблоки? Для ответа на этот вопрос необходимо провести дальнейшую конкретизацию: определить сорт яблок, уточнить магазин, в котором куплены яблоки, и узнать цену этого сорта яблок в этом магазине. Формулировка и решение этой задачи составляет предмет <b>упражнения</b> для пытливого читателя.

#### Задача 5

В задаче 2 нам нужно было сложить <b>два</b> числа. Проведем <b>абстрагирование</b> этой задачи: задан список чисел произвольной длины, необходимо вычислить сумму всех этих чисел. Например, задан список чисел [3, 5, 6, 8], их сумма равна 22, а для списка [4, 7, 12, 2, 5 ,7] сумма равна 37. Подобные задачи уже не слишком удобно решать вручную. Абстрагирование и использование языков программирования помогает нам переложить эту работу на компьютер. Например, на языке Питон, который здесь используется, вычисление сумм чисел из списков записывается в виде

In [5]:
sum([3, 5, 6, 8])

22

и

In [6]:
sum([4, 7, 12, 2, 5, 7])

37

Более того, мы можем пойти по пути абстрагирования дальше, и определить функцию, которая вычисляет сумму всех чисел из заданного списка произвольной длины:

In [7]:
def sum_spisok(spisok):
    """ суммирование всех чисел из списка 'spisok' """
    return sum(spisok)

Теперь вычисление сумм чисел из списков можно производить обращением к функции, например,

In [8]:
spisok = [3, 5, 6, 8]
sum_spisok(spisok)

22

или так:

In [9]:
sum_spisok([1,2,3,4,5,6,7,8,9,10])

55

А вот как выглядит решение знаменитой задачи Гаусса (просуммировать все числа от 1 до 100), которую он решил, будучи учеником младших классов:

In [10]:
sum_spisok(range(1, 101))

5050

Теперь-то мы (пользуясь таким мощным инструментом, как <b>абстракция</b>) знаем, как вычислять сумму членов <b>арифметической прогрессии</b>, а тогда, в XVIII веке, Гаусс догадался сам, что сумма первых $n$ чисел равна

$$
1 + 2 + \cdots + n = \frac{n(n+1)}{2},
$$

что для $n=100$ дает

$$
1 + 2 + \cdots + 100 = \frac{100\cdot 101}2 = 50\cdot 101 = 5050,
$$

как и у Гаусса.